### DISCUSSION TITLE
TypeError: Cannot read properties of undefined (reading 'playNote')

### DISCUSSION THREAD

**Discussion by theslyprofessor** [djipco/webmidi#181] [Q&A]
Upon trying to complete the very first example in the getting started section, I get this error. The midi initialization worked properly on the previous step.

**Comment by theslyprofessor** [comment]
I've since abandoned this idea, and downloaded the dev branch. I was able to get this particular thing to work when using the dev branch and the corresponding code for the dev branch.

**Comment by djipco** [comment]
I'm glad it's working for you and I encourage you to keep posting to this thread if anything goes wrong. I'm actually recruiting people to try out the new version and let me know if they run into troubles.

As a new user, your perspective is very valuable.

Cheers!

**Reply by theslyprofessor** [reply]
Hello, thanks for your response! I am trying to comment on the other thread regarding channels, but commenting seems to be closed. I was actually using webmidi 3 for that and could use a little more clarification.

**Reply by theslyprofessor** [reply]
Ah, ok. I will post here. Running off to school now, will comment soon.

**Reply by djipco** [reply]
> Hello, thanks for your response! I am trying to comment on the other thread regarding channels, but commenting seems to be closed. 

I moved your issue to the discussions section because, usually, issues are reserved for bugs. The discussions section is the preferred place to ask questions about library usage.

**Reply by theslyprofessor** [reply]
Thanks for the clarification and help, greatly appreciated! I was able to make progress, but I am finding the noteOn events confusing in the documentation. How do I grab notes? I am trying this, but it's not working:

```
  function onEnabled() {
    let input = WebMidi.inputs[2];
    input.addListener('noteon', "all", function(e) {
      console.log("Pitch value: " + e.note);
    });
  }
```

substituting e.value in for e.note, as per the documentation, floating point values between 0 and 1 are returned, but I don't know how to interpret that. Looking at the API docs, I saw a "note" payload entry, but I don't know how to access it from the larger payload.

Looking deeper into the documentation, would this help me? It's unclear to me what these types represent, and what the callback function actually does: 

> You can also import the types, if you need them:
```
import WebMidi, { InputEventNoteon, InputEventNoteoff } from "webmidi";
  input.addListener("noteon", "all", (event: InputEventNoteon) => {
 ...
}) 
```

**Reply by theslyprofessor** [reply]
Ah, I think I fixed it with this:
```
  function onEnabled() {
    let input = WebMidi.inputs[2];
    input.addListener('noteon', "all", function(e) {
      console.log("Pitch value: " + e.note.number);
    });
  }
```
I'm still having trouble understanding the use case for the following piece of documentation, and I don't understand how to shape the particular callback function.

`You can also import the types, if you need them:`
```
import WebMidi, { InputEventNoteon, InputEventNoteoff } from "webmidi";
  input.addListener("noteon", "all", (event: InputEventNoteon) => {
 ...
```

**Reply by djipco** [reply]
>```ts
>import WebMidi, { InputEventNoteon, InputEventNoteoff } from "webmidi";
>```

This code applies to using TypeScript in v2. If you are using v3, you should only refer to the [documentation for v3](https://djipco.github.io/webmidi/docs/). By the way, a new website (and documentation) is on its way. It is currently accessible at https://djipco.github.io/webmidi and contains the most up-to-date information about v3.

Once it's fully ready, it will be put online at  https://webmidijs.org
